pyccel/ast/builtins.py

```diff
@@ -297,83 +257,69 @@ def __len__(self):
-        if (self._arg_dtypes is None):
-            raise RuntimeError("This function cannot be used until the type has been infered")
         return self._is_homogeneous
 
-    def set_arg_types(self,d_vars):
-        """ set the types of each argument by providing
-        the list of d_vars calculated using the function
-        _infere_type in parser/semantics.py
-
-        This allows the homogeneity properties to be calculated
-        """
-        self._arg_dtypes = d_vars
-        dtypes = [str(a['datatype']) for a in d_vars]
-
-        #If all arguments are provided then the homogeneity must be checked
-        if (len(self._args)==len(d_vars)):
-            self._is_homogeneous = len(set(dtypes))==1
-
-            if self._is_homogeneous and d_vars[0]['datatype'] == 'tuple':
-                self._is_homogeneous = all(a.is_homogeneous for a in self._vars)
-                if self._is_homogeneous:
-                    dtypes = [str(v.homogeneous_dtype) for v in self._vars]
-                    self._is_homogeneous = len(set(dtypes))==1
-                    if self._is_homogeneous:
-                        self._homogeneous_dtype = d_vars[0].homogeneous_dtype
-            else:
-                self._homogeneous_dtype = d_vars[0]['datatype']
-        else:
-            # If one argument is provided then the tuple must be homogeneous
-            # unless it contains tuples as these tuples are not necessarily homogeneous
-            assert(len(d_vars)==1)
-            self._is_homogeneous = True
-            self._homogeneous_dtype = d_vars[0]['datatype']
-            if d_vars[0]['datatype'] == 'tuple':
-                self._is_homogeneous = self._vars[0]._is_homogeneous
-                if self._is_homogeneous:
-                    self._is_homogeneous_dtype = self._vars[0]._homogeneous_dtype
-
-    @property
-    def arg_types(self):
-        if (self._arg_dtypes is None):
-            raise RuntimeError("This function cannot be used until the type has been infered")
-        return self._arg_dtypes
-
 #==============================================================================
-class Len(Function):
+class Len(Function, PyccelAstNode):
 
     """
     Represents a 'len' expression in the code.
     """
-    is_zero = False
+
+    _rank = 0
+    _shape = ()
+    _precision = default_precision['int']
+    _dtype = NativeInteger()
 
     def __new__(cls, arg):
         return Basic.__new__(cls, arg)
 
-    def __init__(self, arg):
-        assumptions = {'integer': True}
-        ass_copy = assumptions.copy()
-        self._assumptions = StdFactKB(assumptions)
-        self._assumptions._generator = ass_copy
-
     @property
     def arg(self):
         return self._args[0]
 
-    @property
-    def dtype(self):
-        return 'int'
-
-    @property
-    def rank(self):
-        return 0
-
 #==============================================================================
-class List(Tuple):
+class List(Tuple, PyccelAstNode):
```

Comment:
Should this be called PythonList (like PythonTuple) to differentiate from the sympy object?

Reply:
Yes, we need a consistent naming scheme for our AST nodes. I plan on doing that in the near future.